src/Models_Functions/MTV/Compute_M0_T1_OnSPGR.m

```diff
@@ -0,0 +1,79 @@
+function [T1, M0] = Compute_M0_T1_OnSPGR(data, flipAngles, TR, b1Map, roi, verbose)
+%Perform a simple linear-least squares data fit on variable flip angle SPGR data 
+%
+% function [M0, T1] = Compute_M0_T1_OnSPGR(data, flipAngles, TR [, b1Map, roi, verbose])
+% -----------------------------------------------------------
+% INPUTS:
+%   data: width x length x slices x flipAngles matrix
+%   flipAngles: vector of flip angles (in degrees) corresponding to last dimension of 'data'
+%   TR: Repetition time in seconds
+%   b1Map: width x length x slices matrix containing relative flip angle
+%         (i.e. if nominal alpha is 60 and measured alpha is 61, then b1Map = 61/60
+%   roi: width x length x slices binary mask 
+%   verbose: logical - for debugging
+
+
+if ndims(data)<3, data = permute(data(:),[2 3 4 1]); end
+[nX, nY, nZ, nFlip] = size(data);
+nVox = nX*nY*nZ;
+
+if (nargin < 4) || isempty(b1Map)
+    b1Map = ones(nX, nY, nZ);
+end
+
+if nargin<5 || isempty(roi)
+    roi = true(nX, nY, nZ);
+end
+
+if nargin<6
+    verbose = 0;
+end
+ 
+if length(b1Map) ~= length(data(:,:,:,1)), error('B1 size is different from data size'); end
+if ~islogical(roi), roi = roi>0; end
+
+% Reshape data into 2D array so that future steps are simpler
+data = reshape(data,[nVox nFlip])'; % Transpose because MATLAB is column-major
+data = data(:, roi(:));
+nVox = sum(roi(:));
+
+% Large matrix with redundant values will make computations faster,
+% but will also consume more memory
+alpha = deg2rad(flipAngles);
+if isrow(alpha), alpha = alpha'; end
+alpha = repmat(alpha,[1 nVox]) .* repmat(b1Map(roi(:))', [nFlip 1]);
+
+% Do the linear least squares fit
+y = data ./ sin(alpha);
+x = data ./ tan(alpha);
+[fittedSlope, fittedIntercept] = LinLeastSquares(x,y);
+
+% Assign arbitrary T1 value if fitted value in unphysical. 
+% Might be better to set this to NaN so users can identify voxels where fit fails
+arbitraryT1 = 0.000000000000001; % Arbitrary value in case fit fails
```

Comment:
@notZaki  As you say in the comment above, I think assigning it to either NaN (or 0) would be more appropriate. NaN is probably best, however it may cause an issue in displaying the image (which could/should be resolved, i.e. reassigned to a value, at the time of plotting, not here).

Reply:
However, assigning it to NaN here that would then cause an issue below in lines 59 and 60, so these should also only be calculate for cases where `isnan(fittedSlope)` is false, and T1/M0 should be assigned to NaN.

Actually, looking it as a whole now, you should likely not use `arbitraryT1` as a variable (or variable name), since you're assigning it to slopes and intercepts. Simply using NaN should be fine, or renaming it is necessary (e.g. something like `failedFitValue`)

Reply:
If failed fits are assigned a value of NaN, then should the background (and any other masked out region) also be NaNs?
My minor nitpick with NaNs is that they mess with the `Min` and `Max` on the Workspace panel:
![image](https://user-images.githubusercontent.com/9019681/53426449-fcd19180-39b4-11e9-9357-3ceb5d04a0ff.png)

Good catch on `arbitraryT1` not actually being assigned to T1.
I updated it so that it the arbitrary value is applied to M0 and T1 instead of the slope. 
Arbitrary value is zero for now. Can be changed later.

Reply:
Good points. What would could do then (not necessarily now, but open an issue for a future change) would be to set all failed fit values and background to 0 like you did now, but find all the values that are only failed fits and return (or save in the object) that as a binary mask (1 = failed fit, 0 = fit ok).

Reply:
The fact that the `Min` and `Max` show `NaN` is a good thing - fit failures should be rare, and fixable, and you want the user to see them as fast as possible. Fit failures either occur because the algorithm is wrong (in which case the developer needs to fix it) or because a background region was processed (in which case the user needs to fix their mask).

Masked out regions should generally be set to zero. They are not errors.

Reply:
Yeah agreed with @spinicist too. I don't think I've encountered many (any?) failed fits for the VFA method before.

Reply:
I've seen plenty, but they have all been a case of user (me) error!

Reply:
Arbitrary value for failed fits is now NaN.

The VFA method has failed fits if the mask isn't applied, and the example code in `vfa_t1` doesn't load a mask. So that's how I've been testing failed fits.

Reply:
Did the fits actually fail, or did they fall below your cutoffs (slope < 0, intersect < 0). I don't think I'd consider those "failed fits", as noise can cause those, like you said outside the masks. Only in the case where the calculated T1 or M0 return NaN or inf's would I consider those truly failed fits.

Reply:
There are slopes and intercepts with NaNs that are caused when the signal is zero for both flip angles, i.e. linear fit is for line joining `(0,0)` -> `(0,0)`. This happens at the edges of the image.

The cases where `slope < 0` was also set to NaN because that leads to a complex T1.

Should the two cases---`slope==NaN` and `slope<0`---be treated differently?

Reply:
In my opinion, slope < 0 is a "failed fit" because it is unphysical. I appear to disagree with @mathieuboudreau on this though.

Reply:
@spinicist I do see you're point, I'm not very strongly attached to that opinion. I guess I prefer to distinguish between failed fits as in the values weren't mathematically calculable (i.e. NaN or inf results), and the physical cutoffs (e.g. T1 < 0 s or T1 > ~5 s). I typically like to handle these separately. But those could all be classified as "failed fits" from a user standpoint.